recipes_source/mobile_perf.rst

```diff
@@ -0,0 +1,191 @@
+to have an optimized model that will probably have shorter execution time
+(higher performance, lower latency) on the mobile device.
+
+1. Use torch.utils.mobile_optimizer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Torch mobile_optimizer package does several optimizations with the model,
+which will help to conv2d and linear operations.
+It pre-packs model weights in an optimized format and fuses ops above with relu
+if it is the next operation.
+
+::
+
+  from torch.utils.mobile_optimizer import optimize_for_mobile
+  traced_model = torch.jit.load("input_model_path")
+  optimized_model = optimize_for_mobile(traced_model)
+  torch.jit.save(optimized_model, "output_model_path")
+
+2. Fuse operators using ``torch.quantization.fuse_modules``
+Do not be confused that fuse_modules is in the quantization package.
+It works for all types of torch script modules.
+``torch.quantization.fuse_modules`` fuses a list of modules into a single module.
+It fuses only the following sequence of modules:
+
+- Convolution, Batch normalization
+- Convolution, Batch normalization, Relu
+- Convolution, Relu
+- Linear, Relu
+
+This script shows how to fuse Batch Normalization in torchvision mobileNetV2.
+
+::
+
+  import torch
+  import torchvision
+
+  # ConvBNReLU: nn.Sequential[nn.Conv2d, nn.BatchNorm2d, nn.ReLU]
+  # InvertedResidual: nn.Sequential[ConvBnReLU, ConvBNReLU, nn.Conv2d, nn.BatchNorm2d]
+  def fuse_model(model):
+      for idx, m in enumerate(model.modules()):
+          name = m._get_name();
+          if name == 'ConvBNReLU':
+              torch.quantization.fuse_modules(m, ['0', '1'], inplace=True)
+          if name == 'InvertedResidual':
+              for idx in range(len(m.conv)):
+                  if type(m.conv[idx]) == torch.nn.Conv2d:
+                      torch.quantization.fuse_modules(m.conv, [str(idx), str(idx + 1)], inplace=True)
+
+  m = torchvision.models.mobilenet_v2(pretrained=True)
+  m.eval()
+  fuse_model(m)
+  torch.jit.trace(m, torch.rand(1, 3, 224, 224)).save("mobilenetV2-bnfused.pt")
+
+Quantization
+------------
+
+
+3. Try a quantized version of your model
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can find more about PyTorch quantization in
+`the dedicated tutorial <https://pytorch.org/blog/introduction-to-quantization-on-pytorch/>`_.
+
+Quantization of the model not only moves computation to int8,
+but also reduces the size of your model on a disk.
+That size reduction helps to reduce disk read operations during the first load of the model and decreases the amount of RAM.
+Both of those resources can be crucial for the performance of mobile applications.
+
+::
+
+  import torch
+  import torchvision
+
+  supported_qengines = torch.backends.quantized.supported_engines
+  print(supported_qengines)
+  model = torchvision.models.quantization.__dict__['mobilenet_v2'](pretrained=True, quantize=True)
+  torch.backends.quantized.engine='qnnpack'
+  model.eval()
+  script_model = torch.jit.script(model)
```

Comment:
I don't actually know the answer to this: Is it preferable to do quantization before or after TorchScript conversion, or does it matter at all?

Reply:
Currently, pytorch only supports doing quantization prior to scripting. We are working on adding support for quantization after scripting, but it is not part of release 1.6 yet.